### PR TITLE
lib: label temporary files with correct selinux context

### PR DESCRIPTION
**Pull request by pbrezina**
Resolves:
https://github.com/pbrezina/authselect/issues/128

### CODE REVIEW

**Comment by pbrezina** [comment]
Authselect first generates temporary files, then rename them to their proper name under /etc/authselect. This unfortunately creates them with a wrong selinux context. With this patch, correct context is used. You can test it with `ls -lZ /etc/authselect`.

File: src/lib/util/selinux.c [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+    }
+
+    seret = setfscreatecon(default_context);
+    if (seret != 0) {
+        ERROR("Unable to set fscreate selinux context!");
+        ret = EIO;
+        goto done;
+    }
+
+    fd = mkstemp(tmpfile);
```
Comment (jhrozek) [inline src/lib/util/selinux.c]: Normally calls don't set errno to 0 on success, don't you want to set ret = errno only on failure?

File: src/lib/util/selinux.c [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+    ret = errno;
+
+    seret = setfscreatecon(original_context);
```
Comment (jhrozek) [inline src/lib/util/selinux.c]: Why do you call this even if mkstemp fails?

File: src/lib/util/selinux.c [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    if (seret != 0) {
+        ERROR("Unable to restore fscreate selinux context!");
+        ret = EIO;
```
Comment (jhrozek) [inline src/lib/util/selinux.c]: Don't you leak fd here?

File: src/lib/util/selinux.c (hunk 1 above)
Comment (pbrezina) [reply src/lib/util/selinux.c]: Right. I tried to find a nice way to call `setfscreatecon` only once to restore the original context, but perhaps there is not any.

File: src/lib/util/selinux.c (hunk 2 above)
Comment (pbrezina) [reply src/lib/util/selinux.c]: To restore original default selinux context for new files.

File: src/lib/util/selinux.c (hunk 3 above)
Comment (pbrezina) [reply src/lib/util/selinux.c]: Yes, I do.

**Comment by pbrezina** [comment]
Thank you. I pushed fixed patch.

**Comment by doverride** [comment]
Hello and thanks for addressing it. Even though I cannot C, I just wanted to just give some more info just to clarify:

For the sake of example, here is a test to see if the result is as expected (i am using type passwd_file_t for the sake of example:

# Tell selinux that /etc/authselect/nsswitch.conf should have type passwd_file_t:
echo "(filecon "/etc/authselect/nsswitch\.conf" file (system_u object_r passwd_file_t ((s0)(s0))))" > mytest.cil

# load it into the policy db:
semodule -i mytest.cil

# verify that selinux got the message:
matchpathcon /etc/authselect/nsswitch.conf

# rm the existing /etc/authselect and make authselect recreate it
# verify that authselect created it with the type we specified:

ls -alZ /etc/authselect/nsswitch.conf
now running restorecon should not try to change it because its right:
restorecon /etc/authselect/nsswitch.conf

# end of test remove mytest.cil: semodule -r mytest && rm mytest.cil

**Comment by doverride** [comment]
If that does not do it for you, you can mke it  a bit broader (handy also for backups)

echo '(filecon "/etc/authselect/nsswitch\\.conf\\..*" file (system_u object_r passwd_file_t ((s0)(s0))))' >> mytest.cil

**Comment by jhrozek** [comment]
Code-wise LGTM, but I haven't tested the code at all. Maybe it would be nice of @doverride could test some test build?

**Comment by doverride** [comment]
Yes if you point me to a link to a scratch built on koji then i can test it but the example above should be easy enough to use to test it yourself

**Comment by pbrezina** [comment]
@doverride Are you sure it should be `system_u object_r passwd_file_t`? If I call `sudo restorecon -RvF /etc/nsswitch.conf` it is `system_u:object_r:etc_t:s0`.

**Comment by doverride** [comment]
Its just an example to test whether it works as expected

**Comment by doverride** [comment]
if you add the filecon:

```
echo '(filecon "/etc/authselect/nsswitch\.conf\..*" file (system_u object_r passwd_file_t ((s0)(s0))))' >> mytest.cil
semodule -i mytest.cil
```
then restorecon will try to reset it to passwd_file_t

**Comment by pbrezina** [comment]
Right, I understand it now. Thank you.

**Comment by pbrezina** [comment]
I pushed a new version. The original version only relabeled the generated files under /etc/authselect but the symlinks still got wrong context. Now even symlinks (/etc/nsswitch.conf, /etc/pam.d/system-auth...) should have a proper context.

**Comment by doverride** [comment]
How could you tell that the symlinks where mislabeled? My example ignored the symlinks.
Also the symlinks arent a problem. The problem was that authselect created the files in /etc/authselect with random suffixes and then renamed the files to the proper names.

This does not apply to the symlinks. The symlinks arent randomly suffixed and so we can deal with those in the policy.

So i would ignore tthe symlinks and only focus on the files in /etc/authselect

**Comment by doverride** [comment]
You see, we can do a lot of labeling with policy. Its just that we cannot deal with files with random names in many cases. If the name of the file (or symlink) can be anticipated then policy can (and should preferably) be used. You want as little selinux awareness in your code as possible, but enough to make it work.

**Comment by pbrezina** [comment]
@doverride In my test, `/etc/nsswitch.conf` was a symlink to `/etc/authselect/nsswitch.conf`. Its label was unconfined_u:object_r:etc_t. After running restorecon, it was relabeled to `system_u:object_r:etc_t` so I assume the label was not correct.

Here is a scratch build that sets the label for generated files and symlinks:
https://koji.fedoraproject.org/koji/taskinfo?taskID=31939960

If it is not necessary to set the label on symlink even if it does not match the result from calling restorecon, I can remove the code.

The algorithm works as:
1. Use selabel_lookup to obtain the default label for a file without tmp suffix.
2. Use setfscreatecon to set this label on new files
3. Create temporary file
4. Restore original label for setfscreatecon

Same for symlinks.

**Comment by doverride** [comment]
Yes please remove the code for the symlink. The issue you describe is not related. It is because you use restorecon with -F (which is special) I will test the scratch built and report back

**Comment by pbrezina** [comment]
Ok. I removed the code for symlinks and here is a new scratch build: https://koji.fedoraproject.org/koji/taskinfo?taskID=31940117

**Comment by doverride** [comment]
The first one works (disregarded the symlink aspect) I will look at the second built as well

**Comment by pbrezina** [comment]
Thank you.

Pushed to master:
* 9451dd55c43ce704e520dd4d62676466cc68b71e

Pushed to 1.0.x:
* 9c91693833c11af3793a87265108aa8094261d1a

**Pull request closed without merging**